<h1 style="text-align: center;"><a title="Data Science-AIMS-Cmr-2021-22">Chapter 2: Data Reading Methods </h1>

**Instructor:** 

* Rockefeller

# Introduction

Before diving deep into data manipulation and analysis, mastering the art of data ingestion using pandas is pivotal. This chapter will guide you through the essentials of reading data from various sources, ensuring a robust foundation for your data exploration journey.

**Learning Objectives:**

* Introducing different files reading methods
* Developing the ability to query dataframes to answer practical questions
* Updating dataframes based on useful insights

# Reading Methods

Python has become a beloved language for text and file munging due to its simple syntax for
interacting with files, intuitive data structures, and convenient features. Pandas features a number of functions for reading tabular data as a DataFrame object. Let's import it

In [1]:
import pandas as pd

A plethore of  reading method is available in the pandas library.  You can have access to it by typing the `tab` key after typing `pd.`:


>#### <font color=#800080>Q:</font> <a class="anchor" id="Task-1"></a>

* How many of them could you see and at first glance, what could they possibly mean to you?

For sake of the context , we will just focus on the 3 main reading methods

* $\Large pd.read\_csv$


* $\Large pd.read\_excel$


* $\Large pd.read\_clipboard$

### The `pandas.read_csv` method 

Reading a csv file requires passing as argument the path to the actual file. For simplicity,  the file should be saved in the same folder as the notebook you are running but you could also specify the path to it. See below.

Here is the pattern to follow:  

`dataframe_name  =  pd.read_csv( "path/to/the/file.csv" )`

In [ ]:
df_ebola_west_africa = pd.read_csv("/home/rockefeller/Documents/Teachings_n_Talks/Teachings/AIMS_Sen25/Data_prep/Week_01/Module_02/data/ebola_WA.csv")

In [ ]:
"data\ebola_WA.csv"

In [3]:
df_ebola_west_africa=pd.read_csv('data/ebola_WA.csv')

In [2]:
df_ebola_west_africa = pd.read_csv("/home/rockefeller/Documents/Teachings_n_Talks/Teachings/AIMS_Sen25/Data_prep/Week_01/Module_02/data/ebola_WA.csv")
df_ebola_west_africa

,Date,Day,Cases_Guinea,Cases_Liberia,Cases_SierraLeone,Cases_Nigeria
0,1/5/2015,289,2776.0,NaN,10030.0,NaN
1,1/4/2015,288,2775.0,NaN,9780.0,NaN
2,1/3/2015,287,2769.0,8166.0,9722.0,NaN
3,1/2/2015,286,NaN,8157.0,NaN,NaN
4,12/31/2014,284,2730.0,8115.0,9633.0,NaN
...,...,...,...,...,...,...
117,3/27/2014,5,103.0,8.0,6.0,NaN
118,3/26/2014,4,86.0,NaN,NaN,NaN
119,3/25/2014,3,86.0,NaN,NaN,NaN
120,3/24/2014,2,86.0,NaN,NaN,NaN


># <font color=#800080>Task 2:</font> <a class="anchor" id="Task-1"></a>


Dr. Amara stood at the front of the quiet conference room on a gloomy afternoon. 

The projector screen behind him displayed the year 2014, a year that everyone in the room knew was significant for West Africa.


1. In 2014, the Ebola crisis broke out in some of our neighboring countries. But before we talk about the details, can anyone explain the difference between an outbreak and a pandemic?



2. After a pause, Dr. Amara pointed to a map of Africa on the screen and said, **"Ebola is not a new virus, but it spread very quickly in 2014.** Does anyone know where it came from and why it spread so fast that year?



3. After a short discussion, Dr. Amara led the group to the day's task. She said, **"On your computers, you will find a folder called `data`. Inside that folder is a file called `ebola_WA.csv`**. This file contains information about how many Ebola cases were reported in **Nigeria**, **Guinea**, and **Sierra Leone** over time. Please take a moment to make sure that the file is there and then load it using the standard methods that we have learned.


4. A few minutes later, as the data appeared on the screens, Dr. Amara asked, "What strange things or patterns do you see in the data? Let's talk about them."

In [5]:
df_ebola_central_africa=pd.read_csv('data/ebola_CA.csv', sep="-")
df_ebola_central_africa

,Unnamed: 0,Date,Day,Cases_Cameroon,Cases_Gabon,Cases_CAR
0,0,1/5/2015,289,50.0,25.0,242.0
1,1,1/4/2015,288,69.0,NaN,NaN
2,2,1/3/2015,287,45.0,93.0,285.0
3,3,1/2/2015,286,395.0,13.0,157.0
4,4,12/31/2014,284,300.0,69.0,250.0
...,...,...,...,...,...,...
117,117,3/27/2014,5,143.0,NaN,NaN
118,118,3/26/2014,4,327.0,NaN,62.0
119,119,3/25/2014,3,92.0,NaN,135.0
120,120,3/24/2014,2,206.0,69.0,243.0


># <font color=#800080>Task 3:</font> <a class="anchor" id="Task-1"></a>


Dr. Amara paused and spoke with understanding. She said, **"Data analysts are often not involved in collecting data. This means we sometimes work with datasets that have unfamiliar structures or entries."**

1. Dr. Amara switched to a new slide and said, **"I have a new set of historical Ebola data from our colleagues in Central Africa**. Please take a look at the data in the file called `'ebola_CA.csv'` and share your first thoughts."


2. Dr. Amara saw that people were starting to look at the new data. She said, **"Once you've looked through the data, use the same method we learned earlier to put the data into a table. Then, tell us what you see."**

In [ ]:
ebola_data_CA =  pd.read_csv('data/ebola_CA.csv')
ebola_data_CA.head()

,Unnamed: 0,Date,Day,Cases_Cameroon,Cases_Gabon,Cases_CAR
0,0,1/5/2015,289,50.0,25.0,242.0
1,1,1/4/2015,288,69.0,NaN,NaN
2,2,1/3/2015,287,45.0,93.0,285.0
3,3,1/2/2015,286,395.0,13.0,157.0
4,4,12/31/2014,284,300.0,69.0,250.0


You might have guessed it. CSV stands for comma separated values. Other files similar file extensions are also used. See below:

* `","` for the csv file
* `"\t"` for a tsv file 
* or any other character that separates the different variables for an observation

As example of the  `.tsv` file, see the gapminder dataset below. Let's read this dataset where variables are separated by a tabulation key.

In [7]:
import pandas as pd

In [9]:
world_gapminder = pd.read_csv('data/gapminder.tsv' , sep = "\t" )
world_gapminder.head(5)

,country,continent,year,lifeExp,pop,gdpPercap
0,Afghanistan,Asia,1952,28.801,8425333,779.445314
1,Afghanistan,Asia,1957,30.332,9240934,820.853030
2,Afghanistan,Asia,1962,31.997,10267083,853.100710
3,Afghanistan,Asia,1967,34.020,11537966,836.197138
4,Afghanistan,Asia,1972,36.088,13079460,739.981106


###  The `pandas.read_excel` method 

Excel files are also quite common files for saving data. The XLS file extension is used for files saved as Microsoft Excel worksheets. Reading an excel file requires the same process as reading a csv file. The only difference lies in specifying the sheet you want to get access to.

The dataset below comes from the mining industry. It contains mining locations in some African countries both at the region and at the district level. See here a link to an interesting website where you can download the shapefiles of such regions.  https://gadm.org/formats.html 

The code below accesses an excel sheet nameed `ds_salaries`and the sheet `salary` .

In [ ]:
df_salaries_002 = pd.read_excel('data/pro_salaries.xlsx', sheet_name = 'menu')


In [10]:
df_salaries = pd.read_excel('data/pro_salaries.xlsx', sheet_name = 'salary')

df_salaries.head(5)

,work_year,experience_level,employment_type,job_title,salary,salary_currency,salary_in_usd,employee_residence,remote_ratio,company_location,company_size
0,2023,SE,FT,Principal Data Scientist,80000,EUR,85847,ES,100,ES,L
1,2023,MI,CT,ML Engineer,30000,USD,30000,US,100,US,S
2,2023,MI,CT,ML Engineer,25500,USD,25500,US,100,US,S
3,2023,SE,FT,Data Scientist,175000,USD,175000,CA,100,CA,M
4,2023,SE,FT,Data Scientist,120000,USD,120000,CA,100,CA,M


**Just from looking at the dataset, at first glance, can you guess the meaning of some of the columns?**

### The `pandas.read_clipboard` method 

The `pandas.read_clipboard` method is a convenient tool that reads the contents of your clipboard directly into a pandas DataFrame. This method can be quite useful when you want to quickly grab tables or data that you've copied from websites, spreadsheets, or other sources without having to save them as files first.


In [11]:
# lets practice some.  
# Take a look at this website https://www.timeanddate.com/

    #--->  Navigate to Sun, Moon and Space
    # --->  Click on Sun Calculator  
    # --->  Enter the name of a city of your choice in the search box and click on it.


import pandas as pd
my_city_df = pd.read_clipboard()
my_city_df

,Oct,Sunrise,Sunset,Length,Diff.,Start,End,Start.1,End.1,Start.2,End.2,Time,Mil. km
0,1,06:57 ↑ (93°),18:57 ↑ (267°),11:59:40,−0:47,05:46,20:08,06:11,19:43,06:36,19:18,12:57 (72.2°),149.763
1,2,06:57 ↑ (94°),18:56 ↑ (266°),11:58:52,−0:47,05:46,20:07,06:11,19:42,06:36,19:17,12:57 (71.8°),149.719
2,3,06:57 ↑ (94°),18:55 ↑ (266°),11:58:05,−0:47,05:46,20:06,06:11,19:41,06:36,19:17,12:56 (71.4°),149.676
3,4,06:57 ↑ (94°),18:55 ↑ (265°),11:57:18,−0:47,05:46,20:05,06:11,19:41,06:36,19:16,12:56 (71.0°),149.632
4,5,06:57 ↑ (95°),18:54 ↑ (265°),11:56:30,−0:47,05:46,20:05,06:11,19:40,06:36,19:15,12:56 (70.6°),149.589
5,6,06:57 ↑ (95°),18:53 ↑ (265°),11:55:43,−0:47,05:46,20:04,06:11,19:39,06:36,19:15,12:55 (70.2°),149.546
6,7,06:58 ↑ (96°),18:52 ↑ (264°),11:54:56,−0:47,05:46,20:04,06:11,19:39,06:36,19:14,12:55 (69.9°),149.503
7,8,06:58 ↑ (96°),18:52 ↑ (264°),11:54:09,−0:46,05:47,20:03,06:11,19:38,06:36,19:13,12:55 (69.5°),149.461
8,9,06:58 ↑ (96°),18:51 ↑ (263°),11:53:22,−0:46,05:47,20:02,06:11,19:37,06:36,19:13,12:55 (69.1°),149.418
9,10,06:58 ↑ (97°),18:50 ↑ (263°),11:52:35,−0:46,05:47,20:02,06:12,19:37,06:36,19:12,12:54 (68.7°),149.376


In [ ]:
pd.read_

In [12]:
my_city_df.to_csv('data/mbour_data.csv', index=False)

There are several other pandas reading methods. Depending on the files you are dealing with, you might have to choose the convenient method. Other popular methods include `pd.read_json` (for `json` files) , `pd.read_html` (for `html` files) , `pd.read_`

The principle stays the same, which is reading from external sources and casting it into a pandas dataframe for facilitating the analysis. 

##  Shape and datatype

In [13]:
ebola_wa = pd.read_csv('data/ebola_WA.csv')


To get the first n rows and the last n rows

In [7]:
ebola_wa.head(4)

,Date,Day,Cases_Guinea,Cases_Liberia,Cases_SierraLeone,Cases_Nigeria
0,1/5/2015,289,2776.0,NaN,10030.0,NaN
1,1/4/2015,288,2775.0,NaN,9780.0,NaN
2,1/3/2015,287,2769.0,8166.0,9722.0,NaN
3,1/2/2015,286,NaN,8157.0,NaN,NaN


In [8]:
ebola_wa.tail(5)

,Date,Day,Cases_Guinea,Cases_Liberia,Cases_SierraLeone,Cases_Nigeria
117,3/27/2014,5,103.0,8.0,6.0,NaN
118,3/26/2014,4,86.0,NaN,NaN,NaN
119,3/25/2014,3,86.0,NaN,NaN,NaN
120,3/24/2014,2,86.0,NaN,NaN,NaN
121,3/22/2014,0,49.0,NaN,NaN,NaN


A random sample of 3 rows 

In [19]:
ebola_wa.sample(3)

,Date,Day,Cases_Guinea,Cases_Liberia,Cases_SierraLeone,Cases_Nigeria
12,12/9/2014,262,NaN,7797.0,NaN,NaN
103,4/21/2014,30,NaN,34.0,NaN,NaN
83,6/18/2014,88,390.0,NaN,136.0,NaN


To check the shape of the dataframe (rows and columns)

In [10]:
ebola_wa.shape

(122, 6)

To check the datatype of each of the columns

In [24]:
ebola_wa[pd.isna(ebola_wa['Cases_Liberia']) == True]

,Date,Day,Cases_Guinea,Cases_Liberia,Cases_SierraLeone,Cases_Nigeria
0,1/5/2015,289,2776.0,NaN,10030.0,NaN
1,1/4/2015,288,2775.0,NaN,9780.0,NaN
6,12/27/2014,280,2695.0,NaN,9409.0,NaN
8,12/21/2014,273,2597.0,NaN,9004.0,NaN
11,12/14/2014,267,2416.0,NaN,8356.0,NaN
13,12/7/2014,260,2292.0,NaN,7897.0,20.0
15,11/30/2014,253,2164.0,NaN,7312.0,20.0
17,11/23/2014,246,2134.0,NaN,6599.0,20.0
20,11/16/2014,239,1971.0,NaN,6073.0,20.0
22,11/11/2014,234,1919.0,NaN,5586.0,20.0


In [25]:
ebola_wa[pd.isna(ebola_wa['Cases_Liberia']) == True].shape

(39, 6)

In [20]:
ebola_wa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122 entries, 0 to 121
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Date               122 non-null    object 
 1   Day                122 non-null    int64  
 2   Cases_Guinea       93 non-null     float64
 3   Cases_Liberia      83 non-null     float64
 4   Cases_SierraLeone  87 non-null     float64
 5   Cases_Nigeria      38 non-null     float64
dtypes: float64(4), int64(1), object(1)
memory usage: 5.8+ KB


* We can notice that the dataset contains 122 entries with their index ranging from 0 to 121

* The columns  **Cases_Guinea, Cases_Liberia, Cases_SierraLeone** and **Cases_Nigeria** are **float64** type.
Which means they contain floating points.

* The ***Day** column is of **int64** type

* Some of the columns contains missing values:
        122-93 = 29 for the Cases_Guinea
        122-83 = 39 for the Cases_Liberia
        122-87 = 35 for the Cases_SierraLeone
        122-38 = 84 for the Cases_Nigeria

To check the list of columns

In [21]:
ebola_wa.columns

Index(['Date', 'Day', 'Cases_Guinea', 'Cases_Liberia', 'Cases_SierraLeone',
       'Cases_Nigeria'],
      dtype='object')

Let's consider another dataset

In [26]:
world_gapminder = pd.read_csv('data/gapminder.tsv' , sep='\t')
world_gapminder.head(5)

,country,continent,year,lifeExp,pop,gdpPercap
0,Afghanistan,Asia,1952,28.801,8425333,779.445314
1,Afghanistan,Asia,1957,30.332,9240934,820.853030
2,Afghanistan,Asia,1962,31.997,10267083,853.100710
3,Afghanistan,Asia,1967,34.020,11537966,836.197138
4,Afghanistan,Asia,1972,36.088,13079460,739.981106


To check the unique values inside a column

In [28]:
world_gapminder['country'].nunique()

142

In [ ]:
world_gapminder['country'].unique()

In [14]:
world_gapminder['continent'].unique()

array(['Asia', 'Europe', 'Africa', 'Americas', 'Oceania'], dtype=object)

In [29]:
world_gapminder['continent'].value_counts()

continent
Africa      624
Asia        396
Europe      360
Americas    300
Oceania      24
Name: count, dtype: int64

In [32]:
count_continents = pd.DataFrame(world_gapminder['continent'].value_counts())

count_continents.to_csv('data/count_continents.csv', index=False)

To get a subset of the data, let's query the dataframe and get some insights

># <font color=#800080>Task 4:</font> <a class="anchor" id="Task-1"></a>


1. Maria, a public health student, was sitting in a busy cafe in Luanda. She turned to her colleague, João, and asked him, **"Do you remember when Angola's life expectancy was above 40 years old?** For a country, what do you think large value of life expectancy can be an indicator of? Find those years in the dataset and report their corresponding population level.



2. Dr. Kofi and Dr. Amina, two economics professors, were sitting under a large acacia tree, enjoying their afternoon tea. Dr. Kofi looked nostalgic as he asked Dr. Amina, "Which African country do you think had the highest GDP in the 80s?"  Could you guess the name of that country before computing it?? After computing it, what are your thoughts about that?

># <font color=#800080>Task 5:</font> <a class="anchor" id="Task-1"></a>


Look at the gapminder dataset and answer the following questions:

1. **Southern European GDP in the 2000s:**

At a financial conference in Johannesburg, people were talking about how the economy is changing. Sofia, a delegate from Greece, turned to Kwame, a well-known economic analyst from Ghana, and asked him, **"Kwame, do you know which Southern European countries had the highest and lowest GDP in the 2000s?"** 
Hint: Use the Southern European countries listed below.

2. **African Life Expectancy in the 1970s:**

At a health and development conference in Addis Ababa, people were talking about how healthcare has changed as compared to in the past. Fatima, a public health student from Algeria, went up to Zola, a respected demographic analyst from Tanzania, and asked her, **"Zola, can you tell me which African countries had the highest and lowest life expectancy in the 1970s? What do you think caused this?"**


3. **Asian vs. European GDP in the 90s:**


During an international economics conference in Dakar, people were talking about which countries used to be the richest. Ravi, a researcher from India, curiously asked Chijioke, a skilled economic analyst from Cameroon, **"Chijioke, how many Asian countries do you think had a higher GDP than the average European country in the 1990s?"**

In [ ]:
SE_countries = ['Albania', 'Andorra', 'Bosnia and Herzegovina', 'Croatia', 'Gibraltar', 
                   'Greece', 'Italy', 'Kosovo', 'Malta', 'Montenegro', 'North Macedonia', 
                   'Portugal', 'San Marino', 'Serbia', 'Slovenia', 'Spain', 'Vatican City']